File: src/main/javascript/crypto/src/libs/IdentifierAttestation.ts
import {KeyPair} from "./KeyPair";
import {AttestationCrypto} from "./AttestationCrypto";
import {Attestation} from "./Attestation";
import {Validateable} from "./Validateable";
import {logger} from "./utils";
import {DEBUGLEVEL} from "../config";
import {Asn1Der, x500NameData} from "./DerUtility";

export class IdentifierAttestation extends Attestation implements Validateable{
    private crypto: AttestationCrypto;
    static OID_OCTETSTRING = "1.3.6.1.4.1.1466.115.121.1.40";
    // ECDSA with recommended (for use with keccak signing since there is no explicit standard OID for this)
    public static DEFAULT_SIGNING_ALGORITHM = "1.2.840.10045.4.2";
    public static HIDDEN_IDENTIFIER_VERSION = 18;
    public static NFT_VERSION = 19;
    public static HIDDEN_TYPE = "HiddenType";
    public static HIDDEN_IDENTIFIER = "HiddenIdentifier";

    // SEE RFC 2079
    public static LABELED_URI = "1.3.6.1.4.1.250.1.57";
    public static LABELED_URI_LABEL = "labeledURI";

    private type:string;
    private identifier:string;

    constructor() {
        super();
    }

    fromCommitment(commitment: Uint8Array, keys: KeyPair){

        this.subjectKey = keys;
        this.setVersion(IdentifierAttestation.HIDDEN_IDENTIFIER_VERSION);
        this.setSubject("CN=");
        this.setSigningAlgorithm(IdentifierAttestation.DEFAULT_SIGNING_ALGORITHM);

        this.setSubjectPublicKeyInfo(keys);
        this.setCommitment(commitment);

        this.type = IdentifierAttestation.HIDDEN_TYPE;
        this.identifier = IdentifierAttestation.HIDDEN_IDENTIFIER;

        this.setUnlimitedValidity();
    }

    private setUnlimitedValidity(){
        super.setNotValidBefore(Date.now());
        // This is used to indicate unlimited validity, see https://tools.ietf.org/html/rfc5280#section-4.1.2.5
        super.setNotValidAfter(Date.parse('31 Dec 9999 23:59:59'));
    }

    static fromData(identifier: string, type: number, keys: KeyPair, secret: bigint){
        let crypto = new AttestationCrypto();
        let commitment = crypto.makeCommitment(identifier, type, secret);
        let me = new this();
        me.fromCommitment(commitment, keys);
        me.type = type.toString();
        me.identifier = identifier;
        return me;
    }

    static fromLabelAndUrl(label: string, URL: string, keys: KeyPair){

        let me = new this();

        me.subjectKey = keys;
        me.setVersion(IdentifierAttestation.NFT_VERSION);
        me.setSubject(me.makeLabeledURI(label, URL));
        me.setSigningAlgorithm(IdentifierAttestation.DEFAULT_SIGNING_ALGORITHM);
        me.setIssuer("CN=attestation.id");
        me.setSerialNumber(1);

        me.setSubjectPublicKeyInfo(keys);
        me.setUnlimitedValidity();

        me.type = label;
        me.identifier = URL;
        return me;
    }

    /**
     * @param label the label of the URL, similar to what is inside <a>...</a>
     * @param URL the URL itself, similar to what is in <a href="...">, note that
     * it should already be URLencoded therefore not containing space
     */
    private makeLabeledURI(label: string, URL: string): string {

        // DERUTF8String labeledURLValue = new DERUTF8String(URL + " " + label);
        // RDN rdn = new RDN(LABELED_URI, labeledURLValue);
        // return new X500Name(new RDN[] {rdn});
        // https://docs.microsoft.com/ru-ru/windows/win32/seccertenroll/about-introduction-to-asn-1-syntax-and-encoding
        // let type = Asn1Der.encode("OBJECT_ID", IdentifierAttestation.LABELED_URI);
        // let value = Asn1Der.encode("UTF8STRING", URL + " " + label);
        //
        // let rdn = Asn1Der.encode("SEQUENCE_30", type + value);
        // let set = Asn1Der.encode("SET", rdn);
        //
        // return Asn1Der.encode("SEQUENCE_30", set);
        return `${IdentifierAttestation.LABELED_URI_LABEL}="${URL} ${label}"`;
    }

    static fromBytes(bytes: Uint8Array){
        let me = new this();
        me.fromBytes(bytes);
        if (!me.checkValidity()) {
            throw new Error("Could not validate object");
        }
        if (me.getVersion() == IdentifierAttestation.NFT_VERSION) {
            let x500Arr = Asn1Der.parseX500Names(me.getSubject());
            let found = false;
            x500Arr.forEach((item:x500NameData)=>{
                if (found) return;
                if (item.type == IdentifierAttestation.LABELED_URI_LABEL) {
                    found = true;
                    let typeAndIdentifier = item.value.split(' ');
                    if (typeAndIdentifier.length != 2) {
                        throw new Error("LabeledURI values should be separated with space and 2 items only: " + item.value);
                    }
                    me.type = typeAndIdentifier[0];
                    me.identifier = typeAndIdentifier[1];
                }
            })
            if (!found) { throw new Error("Cant find LABELED_URI"); }
        } else {
            me.type = IdentifierAttestation.HIDDEN_TYPE;
            me.identifier = IdentifierAttestation.HIDDEN_IDENTIFIER;
        }

        return me;
    }

    setSubjectPublicKeyInfo(keys: KeyPair){
        this.subjectKey = keys;
    }

    setCommitment(encodedRiddle: Uint8Array) {
        this.commitment = encodedRiddle;
    }

    checkValidity(): boolean {
        if (!super.checkValidity()) {
            logger(DEBUGLEVEL.LOW, "IdentifierAttestation super.checkValidity() filed;")
            return false;
        }
        if (this.getVersion() != IdentifierAttestation.HIDDEN_IDENTIFIER_VERSION && this.getVersion() != IdentifierAttestation.NFT_VERSION) {
            logger(DEBUGLEVEL.LOW, "The version number is " + this.getVersion() + ", it must be either " + IdentifierAttestation.HIDDEN_IDENTIFIER_VERSION + " or " + IdentifierAttestation.NFT_VERSION);
            return false;
        }
        if (this.getSigningAlgorithm() !== IdentifierAttestation.DEFAULT_SIGNING_ALGORITHM) {
            logger(DEBUGLEVEL.LOW, "The subject is supposed to only be an Ethereum address as the Common Name");
            return false;
        }

        if (this.getVersion() == IdentifierAttestation.NFT_VERSION) {
            if (!this.subject.includes(IdentifierAttestation.LABELED_URI_LABEL+"=")) {
                logger(DEBUGLEVEL.LOW, "A NFT Identifier attestation must have a labeled uri as subject. subject = " + this.getSubject());
                return false;
            }
        }

        if (this.getVersion() == IdentifierAttestation.HIDDEN_IDENTIFIER_VERSION) {
            // Ensure that there is a commitment as part of the attestation
            if (this.getCommitment().length < AttestationCrypto.BYTES_IN_DIGEST) {
                logger(DEBUGLEVEL.LOW, "The attestation does not contain a valid commitment");
                return false;
            }
        }

        return true;
    }

    setIssuer(issuer: string){
        this.issuer = issuer;
    }

    public getSerialNumber(): number {
        return this.serialNumber;
    }

    public setSerialNumber(serialNumber: number) {
        this.serialNumber = serialNumber;
    }

    public getAddress(): string {
        // Remove the "CN=" prefix
        return this.subjectKey.getAddress();
    }
}

File: src/main/java/org/tokenscript/attestation/IdentifierAttestation.java
package org.tokenscript.attestation;

import static org.tokenscript.attestation.core.AttestationCrypto.BYTES_IN_DIGEST;

import org.tokenscript.attestation.core.AttestationCrypto;
import org.tokenscript.attestation.core.ExceptionUtil;
import org.tokenscript.attestation.core.SignatureUtility;
import org.tokenscript.attestation.core.URLUtility;
import org.tokenscript.attestation.core.Validateable;
import java.io.IOException;
import java.io.InvalidObjectException;
import java.math.BigInteger;
import java.net.MalformedURLException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.Clock;
import java.util.Date;
import java.util.Locale;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.asn1.ASN1Boolean;
import org.bouncycastle.asn1.ASN1EncodableVector;
import org.bouncycastle.asn1.ASN1ObjectIdentifier;
import org.bouncycastle.asn1.ASN1OctetString;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.DEROctetString;
import org.bouncycastle.asn1.DERSequence;
import org.bouncycastle.asn1.DERUTF8String;
import org.bouncycastle.asn1.x500.RDN;
import org.bouncycastle.asn1.x500.X500Name;
import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.util.PublicKeyFactory;
import org.bouncycastle.crypto.util.SubjectPublicKeyInfoFactory;

public class IdentifierAttestation extends Attestation implements Validateable {
  public enum AttestationType {
    PHONE ("phone"),
    EMAIL ("email"),
    INETPERSONA("InetPersona");

    private final String type;

    AttestationType(String type) {
      this.type = type;
    }

    public boolean equals(String otherType) {
      return type.equals(otherType);
    }

    public String toString() {
      return this.type;
    }

    public static AttestationType getType(String stringType) throws IllegalArgumentException {
      AttestationType type;
      switch (stringType.toLowerCase()) {
        case "mail":
        case "email":
          type = AttestationType.EMAIL;
          break;
        case "phone":
          type = AttestationType.PHONE;
          break;
        case "inetpersona":
          type = AttestationType.INETPERSONA;
          break;
        default:
          System.err.println("Could not parse identifier type, must be either \"mail\", \"phone\" or \"InetPersona\"");
          throw new IllegalArgumentException("Wrong type of identifier");
      }
      return type;
    }
  }

  private static final Logger logger = LogManager.getLogger(IdentifierAttestation.class);
  public static final int HIDDEN_IDENTIFIER_VERSION = 18;
  public static final int NFT_VERSION = 19;
  public static final String HIDDEN_TYPE = "HiddenType";
  public static final String HIDDEN_IDENTIFIER = "HiddenIdentifier";

  // SEE RFC 2079
  public static final ASN1ObjectIdentifier LABELED_URI = new ASN1ObjectIdentifier("1.3.6.1.4.1.250.1.57");
  // ECDSA with recommended (for use with keccak signing since there is no explicit standard OID for this)
  public static final AlgorithmIdentifier DEFAULT_SIGNING_ALGORITHM = new AlgorithmIdentifier(new ASN1ObjectIdentifier("1.2.840.10045.4.2"));


  private final String identifier;
  private final String type;


  /**
   * Constructs a new identifier attestation based on a secret, with unlimited validity by default
   * You still need to set the optional fields, that is
   * issuer, smartcontracts
   */
  public IdentifierAttestation(String identifier, AttestationType type, AsymmetricKeyParameter key, BigInteger secret)  {
    super();
    super.setVersion(HIDDEN_IDENTIFIER_VERSION);
    super.setSubject("CN=");
    super.setSigningAlgorithm(DEFAULT_SIGNING_ALGORITHM);
    try {
      SubjectPublicKeyInfo spki = SubjectPublicKeyInfoFactory.createSubjectPublicKeyInfo(key);
      super.setSubjectPublicKeyInfo(spki);
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode asn1", e);
    }
    setCommitment(AttestationCrypto.makeCommitment(identifier, type, secret));
    setUnlimitedValidity();

    this.identifier = identifier;
    this.type = type.toString();
  }

  /**
   * Restores an attestation based on an already existing commitment, with unlimited validity by default
   * You still need to set the optional fields, that is
   * issuer, smartcontracts
   */
  public IdentifierAttestation(byte[] commitment, AsymmetricKeyParameter key)  {
    super();
    super.setVersion(HIDDEN_IDENTIFIER_VERSION);
    super.setSubject("CN=");
    super.setSigningAlgorithm(DEFAULT_SIGNING_ALGORITHM);
    try {
      SubjectPublicKeyInfo spki = SubjectPublicKeyInfoFactory.createSubjectPublicKeyInfo(key);
      super.setSubjectPublicKeyInfo(spki);
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode asn1", e);
    }
    setCommitment(commitment);
    setUnlimitedValidity();
    this.type = HIDDEN_TYPE;
    this.identifier = HIDDEN_IDENTIFIER;

  }

  /**
   * Constructs an attestation with *public* identifier and with unlimited validity by default
   * You still need to set the optional fields, that is
   * issuer, smartcontracts.
   * This is done using labeledURL, hence URL must be a valid URL
   */
  public IdentifierAttestation(String label, String URL, AsymmetricKeyParameter key) throws MalformedURLException {
    super();
    super.setVersion(NFT_VERSION);
    super.setSubject(makeLabeledURI(label, URL));
    super.setSigningAlgorithm(DEFAULT_SIGNING_ALGORITHM);
    super.setIssuer("CN=attestation.id");
    super.setSerialNumber(1);
    try {
      SubjectPublicKeyInfo spki = SubjectPublicKeyInfoFactory.createSubjectPublicKeyInfo(key);
      super.setSubjectPublicKeyInfo(spki);
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode asn1", e);
    }
    setUnlimitedValidity();
    this.type = label;
    this.identifier = URL;

  }

  public IdentifierAttestation(byte[] derEncoding) throws IOException, IllegalArgumentException {
    super(derEncoding);
    if (!checkValidity()) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not validate object"));
    }
    if (getVersion() == NFT_VERSION) {
      RDN[] labeledURIRDN = (new X500Name(getSubject())).getRDNs(LABELED_URI);
      DERUTF8String labeledURI = (DERUTF8String) labeledURIRDN[0].getFirst().getValue();
      String[] typeAndIdentifier = URLDecoder.decode(labeledURI.getString()).split(" ");
      this.type = typeAndIdentifier[0];
      this.identifier = typeAndIdentifier[1];
    } else {
      this.type = HIDDEN_TYPE;
      this.identifier = HIDDEN_IDENTIFIER;
    }
  }

  /**
   * @param label the label of the URL, similar to what is inside <a>...</a>
   * @param URL the URL itself, similar to what is in <a href="...">, note that
   * it should already be URLencoded therefore not containing space
   */
  private X500Name makeLabeledURI(String label, String URL)  {
    DERUTF8String labeledURLValue = new DERUTF8String(URL + " " + label);
    RDN rdn = new RDN(LABELED_URI, labeledURLValue);
    return new X500Name(new RDN[] {rdn});
  }

  private void setUnlimitedValidity() {
    try {
      super.setNotValidBefore(new Date(Clock.systemUTC().millis()));
      SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMddHHmmss'Z'", Locale.US);
      // This is used to indicate unlimited validity, see https://tools.ietf.org/html/rfc5280#section-4.1.2.5
      Date notValidAfter = dateFormat.parse("99991231235959Z");
      super.setNotValidAfter(notValidAfter);
    } catch (ParseException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not parse date", e);
    }
  }

  /**
   * Verifies that the the attestation is in fact a valid identifier attestation, in relation to field values.
   * @return true if the field values reflect that this is a standard attestation
   */
  @Override
  public boolean checkValidity() {
    if (!super.checkValidity()) {
      logger.error("Could not check validity of the underlying attestation");
      return false;
    }
    if (getVersion() != HIDDEN_IDENTIFIER_VERSION && getVersion() != NFT_VERSION) {
      logger.error("The version number is " + getVersion() + ", it must be either " + HIDDEN_IDENTIFIER_VERSION + " or " + NFT_VERSION);
      return false;
    }
    if (!getSigningAlgorithm().equals(DEFAULT_SIGNING_ALGORITHM)) {
      logger.error("The subject is supposed to only be an Ethereum address as the Common Name");
      return false;
    }
    if (getVersion() == NFT_VERSION) {
      String subject = getSubject();
      if (!subject.contains(LABELED_URI.getId())) {
        logger.error("A NFT Identifier attestation must have a labeled uri as subject");
        return false;
      }
    }
    if (getVersion() == HIDDEN_IDENTIFIER_VERSION) {
      // Ensure that there is a commitment as part of the attestation
      try {
        if (getCommitment().length < BYTES_IN_DIGEST) {
          logger.error("The attestation does not contain a valid commitment");
          return false;
        }
      } catch (Exception e) {
        logger.error("It was not possible to decode the attestation commitment");
        return false;
      }
    }
    return true;
  }

  public byte[] getCommitment() {
    // Need to decode twice since the standard ASN1 encodes the octet string in an octet string
    ASN1Sequence extensions = DERSequence.getInstance(getExtensions().getObjectAt(0));
    // Index in the second DER sequence is 2 since the third object in an extension is the actual value
    return ASN1OctetString.getInstance(extensions.getObjectAt(2)).getOctets();
  }

  public String getAddress() {
    try {
      return SignatureUtility.addressFromKey(PublicKeyFactory.createKey(getSubjectPublicKeyInfo()));
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not decode the address", e);
    }
  }

  /**
   * Set a commitment and sets it as an Attribute on the Attestation/
   * @return A proof of knowledge of the riddle
   */
  private void setCommitment(byte[] encodedRiddle) {
    ASN1EncodableVector extensions = new ASN1EncodableVector();
    extensions.add(Attestation.OID_OCTETSTRING);
    extensions.add(ASN1Boolean.TRUE);
    extensions.add(new DEROctetString(encodedRiddle));
    // Double Sequence is needed to be compatible with X509V3
    this.setExtensions(new DERSequence(new DERSequence(extensions)));
  }

  public String getAsUrlWithIdentifier() {
    String encodedIdentifier = URLEncoder.encode(this.identifier, StandardCharsets.UTF_8);
    String encodedType = URLEncoder.encode(this.type, StandardCharsets.UTF_8);
    return getAsUrlWithoutIdentifier() + "&" + encodedType + "=" + encodedIdentifier;
  }

  public String getAsUrlWithoutIdentifier() {
    try {
      return URLUtility.encodeData(getDerEncoding());
    } catch (InvalidObjectException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not get DER encoding", e);
    }
  }

  @Override
  public byte[] getDerEncoding() throws InvalidObjectException {
    return super.getDerEncoding();
  }

  @Override
  public byte[] getPrehash() {
    return super.getPrehash();
  }

  @Override
  public void setVersion(int version) {
    throw ExceptionUtil.throwException(logger,
        new RuntimeException("Not allowed to be manually set in concrete Attestation"));
  }

  @Override
  public void setSubject(String subject) {
    throw ExceptionUtil.throwException(logger,
        new RuntimeException("Not allowed to be manually set in concrete Attestation"));
  }
}
